MHVLib/Authentication/MHVShellAuthService.m

```diff
@@ -0,0 +1,222 @@
+                    
+                    self.isAuthInProgress = NO;
+                    
+                    return;
+                }
+                
+                NSString *instanceId = [self InstanceIdFromUrl:successUrl];
+                
+                if ([NSString isNilOrEmpty:instanceId])
+                {
+                    completion (nil, [NSError error:[NSError MHVUnknownError] withDescription:@"Failed to obtain an instanceId from the authorization service."]);
+                    
+                    self.isAuthInProgress = NO;
+                    
+                    return;
+                }
+                
+                completion(instanceId, nil);
+                
+                self.isAuthInProgress = NO;
+            });
+        }];
+    });
+}
+
+- (void)authorizeAdditionalRecordsWithViewController:(UIViewController *_Nullable)viewController
+                                            shellUrl:(NSURL *)shellUrl
+                                         masterAppId:(NSUUID *)masterAppId
+                                          completion:(void (^_Nullable)(NSError *_Nullable error))completion
+{
+    MHVASSERT_PARAMETER(shellUrl);
+    MHVASSERT_PARAMETER(masterAppId);
+    
+    dispatch_async(self.authQueue, ^
+    {
+        if (!completion)
+        {
+            return;
+        }
+        
+        if (self.isAuthInProgress)
+        {
+            completion([NSError error:[NSError MHVOperationCannotBePerformed] withDescription:@"Another authentication operation is currenlty running."]);
+            
+            return;
+        }
+        
+        self.isAuthInProgress = YES;
+        
+        if (!shellUrl || !masterAppId)
+        {
+            if (completion)
+            {
+                completion([NSError error:[NSError MVHInvalidParameter] withDescription:@"One or more required parameters are missing."]);
+            }
+        }
+        
+        NSString *queryString = [NSString stringWithFormat:kRecordAuthQueryFormat, masterAppId, [self mraBooleanString]];
+        
+        queryString = [queryString stringByAddingPercentEncodingWithAllowedCharacters:NSCharacterSet.URLPathAllowedCharacterSet];
+        
+        [self authenticateInBrowserWithViewController:viewController
+                                             shellUrl:shellUrl
+                                          queryString:queryString
+                                           completion:^(NSURL * _Nullable successUrl, NSError * _Nullable error)
+        {
+            dispatch_async(self.authQueue, ^
+            {
+                if (completion)
+                {
+                    completion(error);
+                }
+                
+                self.isAuthInProgress = NO;
+            });
+        }];
+    });
+}
+
+#pragma mark - Private
+
+- (NSString *)mraBooleanString
+{
+    return self.configuration.isMultiRecordApp ? @"true" : @"false";
+}
+
+- (NSString *)miaParamString
```

Comment:
The initialism actually stands for Multi-Instance Aware. I'm going to leave this as is for now, until we have something working to unblock everyone else. We can come back and make improvements later.